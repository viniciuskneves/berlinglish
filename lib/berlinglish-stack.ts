import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { Rule, Schedule } from "aws-cdk-lib/aws-events";
import { LambdaFunction } from "aws-cdk-lib/aws-events-targets";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

export class BerlinglishStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const lambdaFunction = new NodejsFunction(
      this,
      "BerlinglishLambdaFunction",
      {
        functionName: "Berlinglish",
        entry: "./lambda/index.ts",
        // bundling: {
        //   format: OutputFormat.ESM,
        // },
        timeout: Duration.seconds(10), // Default 3 seconds timeout is not enough to do all the process, it takes around 5 seconds
        environment: {
          TWITTER_APP_KEY: process.env.TWITTER_APP_KEY!,
          TWITTER_APP_SECRET: process.env.TWITTER_APP_SECRET!,
          TWITTER_ACCESS_TOKEN: process.env.TWITTER_ACCESS_TOKEN!,
          TWITTER_ACCESS_SECRET: process.env.TWITTER_ACCESS_SECRET!,
        },
      }
    );

    new Rule(this, "BerlinglishCron", {
      ruleName: "Berlinglish",
      schedule: Schedule.cron({
        minute: "0",
        hour: "7,9,13,15", // Trying 4x a day. It runs UTC so currently CEST is UTC+2 --> 9,11,15,17
      }),
      targets: [new LambdaFunction(lambdaFunction)],
    });
  }
}
