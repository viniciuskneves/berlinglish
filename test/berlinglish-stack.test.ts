import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { BerlinglishStack } from "../lib/berlinglish-stack";

test.skip("BerlinglishStack", () => {
  const app = new cdk.App();
  // WHEN
  const stack = new BerlinglishStack(app, "MyTestStack");
  // THEN
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
