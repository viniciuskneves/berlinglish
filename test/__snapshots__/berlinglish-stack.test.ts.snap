// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BerlinglishStack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BerlinglishCron69625905": Object {
      "Properties": Object {
        "Name": "Berlinglish",
        "ScheduleExpression": "cron(0 7,9,13,15 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "BerlinglishLambdaFunctionE3FE89FF",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "BerlinglishCronAllowEventRuleMyTestStackBerlinglishLambdaFunctionCB65A1425BA3EC46": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "BerlinglishLambdaFunctionE3FE89FF",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "BerlinglishCron69625905",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "BerlinglishLambdaFunctionE3FE89FF": Object {
      "DependsOn": Array [
        "BerlinglishLambdaFunctionServiceRole40C78245",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "03a44088f4edbc2c236aaa4ea01f633770f4f92bd39709103864655b3134043e.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TWITTER_ACCESS_SECRET": "wFs4DNH8fYC1wicq3CVQzOohMGTjcbiWXvW0WRdTkdHKB",
            "TWITTER_ACCESS_TOKEN": "1245073093242040320-8Sqlav7odp6w0WtP6BkXEH16wjGMpU",
            "TWITTER_APP_KEY": "4eVcI3LiaPr3KqQlBGqjck1an",
            "TWITTER_APP_SECRET": "gAfLDGs5Nbye2rROKSQYqaoMNOgTLfOVqQboSTQ3EOTOgwys69",
          },
        },
        "FunctionName": "Berlinglish",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BerlinglishLambdaFunctionServiceRole40C78245",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 10,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BerlinglishLambdaFunctionServiceRole40C78245": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
